API Profile Support

derekbekoe opened the pull request:
- Use published SDKs available on PyPI now
- Loading versioned models for network and resources, object model initialization changes for 2015-* profile (#2812)
- Custom ca certs - Setting Requests_ca_bundle environment variable (#2813)
- changes for 2015-* profile to work against azure stack  (#2794)
- Loading versioned Resources client and versioned models in VM
- Updating the supported api-versions for the 2015-example profile
- Fixing network, compute sdk load errors for 2015-* profiles (+5 squashed commit)
- Support multi-API versioned Storage dataplane SDK (#2796)
- Support multi-versioned mgmt SDK (#2526)
- Fix token "management" endpoint is being used in the place of "activeDirectoryResourceId" (#2410)
- Add profile switching params and profile listing command (#2398)
- Use ARM 'resource manager' endpoint if ASM 'management' endpoint not set

Closes https://github.com/Azure/azure-cli/issues/2578
Closes https://github.com/Azure/azure-cli/issues/2579
Closes https://github.com/Azure/azure-cli/issues/2577


---

This checklist is used to make sure that common guidelines for a pull request are followed.

### General Guidelines

- [X] The PR has modified HISTORY.rst with an appropriate description of the change (see [Modifying change log](https://github.com/Azure/azure-cli/tree/master/doc/authoring_command_modules#modify-change-log)).

### Command Guidelines

- [X] Each command and parameter has a meaningful description.
- [X] Each new command has a test.

(see [Authoring Command Modules](https://github.com/Azure/azure-cli/tree/master/doc/authoring_command_modules))

tjprescott commented on src/azure-cli-core/azure/cli/core/cloud.py:
```diff
@@ -67,6 +68,13 @@ def __init__(self,  # pylint: disable=too-many-arguments
 
+    def has_endpoint_set(self, endpoint_name):
+        try:
+            getattr(self, endpoint_name)
```
Can't this just be simplified to
```Python
return hasattr(self, endpoint_name)
```

derekbekoe replied:
The behavior for `hasattr` for [Python 2](https://docs.python.org/2/library/functions.html#hasattr) vs [Python 3](https://docs.python.org/3/library/functions.html#hasattr) is *slightly* different.
We actually want the Python 2 implementation.
I'll add a comment in the code to clarify.

tjprescott commented on src/azure-cli-core/azure/cli/core/cloud.py:
```diff
@@ -106,12 +114,23 @@ def __init__(self,
         self.is_active = is_active
 
+    def __str__(self):
+        o = {}
```
Could be simplfied:
```Python
o = {
  'profile': self.profile,
  ...
}
```

derekbekoe replied:
done

tjprescott commented on src/azure-cli-core/azure/cli/core/commands/parameters.py:
```diff
@@ -96,9 +95,16 @@ def __contains__(self, other):
 
+def model_choice_list(resource_type, model_name):
+    model = get_versioned_models(resource_type, model_name)
+    return enum_choice_list(model) if model else {}
```
Should this return [] as the else since it's an choice list?

derekbekoe replied:
`enum_choice_list` returns an empty dict so it made sense to do the same here.
Actually since the usage is `**model_choice_list`, it has to be a dict, not a list.
```
$ python
>>> val = {'n': 1}
>>> 
>>> def f(**kwargs):
...     print(kwargs)
... 
>>> f(**val)
{'n': 1}
>>> 
>>> l = []
>>> f(**l)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: f() argument after ** must be a mapping, not list
>>>
```

tjprescott commented on src/azure-cli-core/azure/cli/core/commands/parameters.py:
```diff
@@ -113,6 +119,14 @@ def _type(value):
+    try:
+        return getattr(mod, enum_val_name).value
+    except AttributeError:
+        return None
```
I'm still concerned about trying to debug a mysterious typo with all of these things being strings. If we were to issue a logger.error, how much spam would we get in --debug for this?

derekbekoe replied:
Hmm yes could do that.
Does this work?
`logger.debug('Unable to get param default %s.%s for %s', enum_name, enum_val_name, resource_type)`
In logs, it would look like:
`Unable to get param default Kind.storage for ResourceType.MGMT_STORAGE`

derekbekoe replied:
Or `Skipping param default Kind.storage for ResourceType.MGMT_STORAGE`

tjprescott replied:
That's fine--my main question is, does this introduce a ton of spam to --debug?  I guess it wouldn't if you were mainly developing against latest, in which case, it would be useful.

derekbekoe replied:
Currently, it only appears once. Don't expect it to be too much.

tjprescott commented on src/azure-cli-core/azure/cli/core/profiles/shared.py:
```diff
@@ -0,0 +1,158 @@
+        raise APIVersionException(resource_type, api_profile)
+
+
+def get_client_class(resource_type):
```
For this change, exposing the minimum public surface area is important. Right now these all show up as public. Can we update the ones that we don't intend to have used by command authors to use the leading _?

derekbekoe replied:
All the methods/attributes/properties in this file are public.

tjprescott replied:
IMO there are way too many public methods then. When people ask "how do I support multiple versions", I would have no idea what to tell them other than to ask you.

derekbekoe replied:
This whole `shared.py` file should not be used by anyone else.
`azure/cli/core/profiles/__init__.py` is what will be used by others and that has the following public methods: `get_api_version`, `get_sdk`, `get_versioned_sdk_path` each of which has their use.

I have renamed `shared.py` to `_shared.py` to make this clearer.

tjprescott commented on src/command_modules/azure-cli-batch/azure/cli/command_modules/batch/custom.py:
```diff
@@ -111,6 +109,7 @@ def update_application(client, resource_group_name, account_name, application_id
 
 def _upload_package_blob(package_file, url):
     """Upload the location file to storage url provided by autostorage"""
+    BlockBlobService = get_sdk_attr('azure.multiapi.storage.blob#BlockBlobService')
```
Couldn't we just get a reference to the versioned SDK and then get this like normal?

derekbekoe replied:
`azure.multiapi.storage.blob` doesn't actually exist.
`azure.multiapi.storage.vYYYY_MM_DD.blob` is what exists and the helper method extracts that.

tjprescott replied:
Right, so why can't I just have one `get_versioned_sdk('azure.blah.blah.blah')` instead of all these other methods that returns 'azure.blah.blah.vYYYY_MM_DD'?

derekbekoe replied:
Ah I see. I've implemented this now whilst unifying get_sdk_attr and get_versioned_models.

tjprescott commented on src/command_modules/azure-cli-find/tests/test_find.py:
```diff
@@ -52,10 +52,10 @@ def test_search_index(self):
             self,
             execute('find -q keyvault list --reindex'),
-            'az keyvault certificate list-versions')
+            'az keyvault list')
```
Why this change?

derekbekoe replied:
This was due to https://github.com/Azure/azure-cli/issues/2805.
Now that's fixed, I should be able to remove this :)

tjprescott commented on src/command_modules/azure-cli-iot/azure/cli/command_modules/iot/_factory.py:
```diff
@@ -9,6 +9,6 @@ def iot_hub_service_factory(_):
 
 def resource_service_factory(**_):
-    from azure.mgmt.resource.resources import ResourceManagementClient
+    from azure.mgmt.resource import ResourceManagementClient
```
I thought this is a definite no-no with profile enabled CLI since it would always get latest (which may more may not be supported on the given cloud)?

derekbekoe replied:
We haven't versioned all modules, iot being one of them.
So yeah it would get the latest in this case which is what we want.

tjprescott commented on src/command_modules/azure-cli-network/azure/cli/command_modules/network/commands.py:
```diff
@@ -31,7 +34,10 @@
 cli_command(__name__, 'network application-gateway stop', 'azure.mgmt.network.operations.application_gateways_operations#ApplicationGatewaysOperations.stop', cf_application_gateways)
-cli_command(__name__, 'network application-gateway show-backend-health', 'azure.mgmt.network.operations.application_gateways_operations#ApplicationGatewaysOperations.backend_health', cf_application_gateways)
+
+if get_api_version(ResourceType.MGMT_NETWORK) in ['2016-09-01']:
```
This mechanism seems very burdensome. I doubt this command is going to vanish in APIs moving forward, and I shouldn't have to add to this (and every other) list for every new API version.  This mechanism needs to be improved with some kinds of convenience methods like "api_version_greater_than" or something.

derekbekoe replied:
Looking into if API versions have an ordering.

tjprescott replied:
API versions are all dates... ?

derekbekoe replied:
Batch has an API version of 2017-01-01.4.0 and then there are preview APIs, and some preview APIs are actually GA, etc.

tjprescott replied:
In the case of batch that definitely still looks like something that can be ordered. In the case of preview releases, there is still a heuristic that can be applied to suggest order. 2017-01-01preview would logically come before 2017-01-01 if both existed.

For example, to support network version 2016-12-01 (which is on my plate to do in the near future), I would have to find everywhere these "api_versions in ['2016-09-01']" type statements exist and add '2016-12-01' to them. If at any point I missed one (or it differently slightly so that find wouldn't catch it) now commands will suddenly disappear by virtue of upping the profile. These are exactly the kind of bugs that having "api_greater_than(...)" would prevent from happening.

tjprescott commented on src/command_modules/azure-cli-storage/azure/cli/command_modules/storage/_factory.py:
```diff
@@ -99,7 +102,7 @@ def queue_data_service_factory(kwargs):
 
 def cloud_storage_account_service_factory(kwargs):
-    from azure.storage import CloudStorageAccount
+    CloudStorageAccount = get_sdk_attr('azure.multiapi.storage#CloudStorageAccount')
```
How is "get_sdk_attr" different from "get_versioned_models"??  It seems like you can do the same thing with either.

derekbekoe replied:
Yes they are similar. Looking at combining them into one.

derekbekoe replied:
Done.

tjprescott commented on src/command_modules/azure-cli-storage/azure/cli/command_modules/storage/_params.py:
```diff
@@ -42,6 +35,25 @@
+TablePayloadFormat = get_sdk_attr('azure.multiapi.storage.table#TablePayloadFormat')
+QueueService = get_sdk_attr('azure.multiapi.storage.queue#QueueService')
+QueuePermissions = get_sdk_attr('azure.multiapi.storage.queue.models#QueuePermissions')
+PublicAccess = get_sdk_attr('azure.multiapi.storage.blob.models#PublicAccess')
```
Why are you not using the "get_versioned_models" to eliminate all these copy-paste redundancy?

derekbekoe replied:
Addressed.

tjprescott commented on src/command_modules/azure-cli-storage/azure/cli/command_modules/storage/_params.py:
```diff
@@ -252,17 +264,21 @@ def register_blob_source_uri_arguments(scope):
-    register_cli_argument('storage account {}'.format(item), 'encryption', nargs='+', help='Specifies which service(s) to encrypt.', validator=validate_encryption, **enum_choice_list(list(EncryptionServices._attribute_map.keys())))  # pylint: disable=protected-access
+    register_cli_argument('storage account {}'.format(item), 'sku', help='The storage account SKU.', **model_choice_list(ResourceType.MGMT_STORAGE, 'SkuName'))
+    es_model = get_versioned_models(ResourceType.MGMT_STORAGE, 'EncryptionServices')
+    if es_model:
```
This seems like an implicit way of saying "if API version  > blah".  I would highly recommend we be specific here and call out API version explicitly. 

Also, why are we using "get_versioned_models" instead of "get_sdk_attr"?

derekbekoe replied:
Done.

tjprescott commented on src/command_modules/azure-cli-storage/azure/cli/command_modules/storage/commands.py:
```diff
@@ -51,17 +54,24 @@ def _dont_fail_not_exist(ex):
+
+if get_api_version(ResourceType.MGMT_STORAGE) in ['2015-06-15']:
+    cli_command(__name__, 'storage account create', custom_path + 'create_storage_account_with_account_type')
+else:
```
So any API versions lower or greater than this should use the newer method?

derekbekoe replied:
There aren't versions lower than 2015-06-15 so it'd be only newer API versions.

tjprescott commented on src/command_modules/azure-cli-vm/azure/cli/command_modules/vm/commands.py:
```diff
@@ -237,36 +242,37 @@ def transform_av_set_collection_output(av_sets):
-cli_command(__name__, 'image show', mgmt_path.format(op_var, op_class, 'get'), cf_images, exception_handler=empty_on_404)
-cli_command(__name__, 'image delete', mgmt_path.format(op_var, op_class, 'delete'), cf_images)
+
+if get_api_version(ResourceType.MGMT_COMPUTE) in ['2016-04-30-preview']:
```
What about VM and VMSS create?  Managed disks didn't exist before this API version. I can't image those commands would work as is.  I don't see different versions of those commands like Storage Account create.

derekbekoe replied:
VM and VMSS create don't fully support older API versions.
This would be part of https://github.com/Azure/azure-cli/issues/2272.

derekbekoe commented on src/command_modules/azure-cli-vm/azure/cli/command_modules/vm/_template_builder.py:
```diff
@@ -698,7 +701,6 @@ def build_vmss_resource(name, naming_prefix, location, tags, overprovision, upgr
     # Build VMSS
     vmss_properties = {
         'overprovision': overprovision,
-        'singlePlacementGroup': single_placement_group,
```
done. thanks.

tjprescott replied:
I thought this PR was not intended to make VM/VMSS create "profile compliant"?

yugangw-msft replied:
I am fine with different options to address it. But since this PR aims to get into master, then we should not just drop the property, as this clearly makes creating large scale-set no longer work.

tjprescott replied:
Oh, yes, definitely. great catch!

derekbekoe commented:
@tjprescott Can you take another look?
I've addressed all comments apart from creating convenience methods like "api_version_greater_than" as we need to get clarity on API version ordering.

derekbekoe commented:
> Main concerns:
> 1. The api conditional check needs to be improved.
> 2. I don't understand when it is appropriate to use get_sdk_attr and get_versioned_models. They seem to do the same thing and it seems they are used interchangeably.
> 3. How are we testing this?? Are we to expect that once merged, these services will work correctly against Stack and other clouds?

1 requires an api ordering which is currently not known.
2. Fixed this. There is not a single `get_sdk` function.
3. It's not expected to completely work on Stack or other clouds yet.

tjprescott commented:
I'm still struggling with these issues:
1. The surface area of the versioning APIs.  I see the following:
`profiles.__init__ has get_api_verison, get_sdk, get_versioned_sdk_path`
`profiles._shared has get_api_version, get_client_class, get_versioned_sdk_path, get_versioned_sdk`
So if _shared is all internal, okay, cool. What's the difference between get_sdk and get_versioned_sdk_path?
2. I don't think merging this without a solution for "API version > blah" makes sense. To me this seems like a bare minimum requirement to have maintainable versioning APIs for people to code against.

troydai commented on src/azure-cli-core/azure/cli/core/_profile.py:
```diff
@@ -66,9 +65,7 @@ def _authentication_context_factory(authority, cache):
-logger.debug("Current active cloud '%s'", CLOUD.name)
-logger.debug(pformat(vars(CLOUD.endpoints)))
-logger.debug(pformat(vars(CLOUD.suffixes)))
+logger.debug(str(CLOUD))
```
Unless the to string conversion includes readable explanation to what it is, I would suggesting print a more readable message like: 
``` PYTHON
logger.debug('Current cloud is {}'.format(str(CLOUD)))
```

derekbekoe replied:
It looks like this:
```
{'endpoints': {'active_directory': 'https://login.microsoftonline.com',
               'active_directory_graph_resource_id': 'https://graph.windows.net/',
               'active_directory_resource_id': 'https://management.core.windows.net/',
               'batch_resource_id': None,
               'gallery': 'https://gallery.azure.com/',
               'management': 'https://management.core.windows.net/',
               'resource_manager': 'https://management.azure.com/',
               'sql_management': 'https://management.core.windows.net:8443/'},
 'is_active': True,
 'name': 'AzureCloud',
 'profile': '2016-sample',
 'suffixes': {'azure_datalake_analytics_catalog_and_job_endpoint': 'azuredatalakeanalytics.net',
              'azure_datalake_store_file_system_endpoint': 'azuredatalakestore.net',
              'keyvault_dns': '.vault.azure.net',
              'sql_server_hostname': '.database.windows.net',
              'storage_endpoint': 'core.windows.net'}}
```

I've added a message though so at least it's possible to search for the string to find which line of code creates that log message.

derekbekoe commented on src/azure-cli-core/azure/cli/core/cloud.py:
```diff
@@ -225,10 +248,19 @@ def get_clouds():
                 setattr(c.suffixes, option.replace('suffix_', ''), config.get(section, option))
+        if c.profile is None:
+            # If profile isn't set, use latest
+            c.profile = 'latest'  # pylint: disable=redefined-variable-type
```
yes that works. thanks for the suggestion!

derekbekoe commented:
@tjprescott 
I have removed `get_versioned_sdk_path` and inlined it to the only place it was used.
Also, see the commit below for the changes to support specifying API version ranges (currently just supports the basic date format as that's all we have right now).
https://github.com/Azure/azure-cli/pull/2834/commits/b84f60e1d17d069633f85be4c483af650a200b45

derekbekoe commented:
merged 🎉 🍾
